### PR TITLE
attention block for obs assimilation

### PR DESCRIPTION
**Pull request by bnb32**

### CODE REVIEW

File: phygnn/layers/custom_layers.py [hunk 1]
```diff
@@ -87,6 +87,475 @@ def call(self, x):
...
+            'strides': patch_size,
+            'padding': 'valid',
+        }
+        pool = (
+            tf.keras.layers.AveragePooling3D(**kwargs)
+            if len(x.shape) == 5
+            else tf.keras.layers.AveragePooling2D(**kwargs)
+        )
+
+        pos = tf.range(x.shape[dim_index]) / x.shape[dim_index]
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py]: i'd prefer linspace for closed boundaries but this is fine too.

so we're going with relative positional encodings right? Have you thought about position 0-1 in each small training example vs. 0-1 in the inference space vs. 0-1 in each fwp chunk? Even if you have a plan we'll have to verify this works. I think there could be some issues around your encoding frequencies working well with the training sample size in 0-1 but poorly when you have a much larger inference size also in 0-1 because the spatial resolution that the transformer is seeing effectively changes. Not confident about this but i think it's something we might have to experiment with.

File: phygnn/layers/custom_layers.py [hunk 2]
```diff
@@ -87,6 +87,475 @@ def call(self, x):
...
+            logger.error(msg)
+            raise ValueError(msg)
+
+        # if input is sparse then tokenize only the non-NaN values with a dense
+        # layer
+        mask = tf.math.is_nan(x)
+        if tf.math.reduce_any(mask):
+            x_tok = x[tf.math.logical_not(mask)]
+            x_tok = tf.reshape(x_tok, (x.shape[0], -1, x.shape[-1]))
+            return tf.keras.layers.Dense(embed_dim)(x_tok)
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py]: This is how long it's been since i've written tensorflow, but i'm wondering if this is an okay way to tokenize because wouldn't it create a new Dense or Conv layer every tokenization call (every batch?)? The weights for these tokenization methods wouldn't get tracked, would they? 

Also, I feel like i might remember it being typical to not use bias terms in the tokenization but i'm just now realizing i've been inconsistent about this and i'm not sure.

File: phygnn/layers/custom_layers.py [hunk 3]
```diff
@@ -87,6 +87,475 @@ def call(self, x):
...
+            Dimension of the tokenized inputs.
+        num_heads : int
+            Number of attention heads
+        key_dim : int
+            Size of each attention head
+        dropout : float
+            Dropout rate for attention weights. Default is 0 (no dropout).
+        patch_size : int
+            Height, width, and depth of patches. Default is 1 for pixel-wise
+            tokenization.
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py]: Patch size of Sup3r hypercube right? not of key/value obs?

File: phygnn/layers/custom_layers.py [hunk 4]
```diff
@@ -87,6 +87,475 @@ def call(self, x):
...
+        )
+        x_tok = tokenizer(x)
+        return tf.reshape(x_tok, (x.shape[0], -1, embed_dim))
+
+
+class Sup3rCrossAttention(TokenizeEncodeBase):
+    """Custom layer to implement cross attention with tokenization and
+    positional encoding. This is typically used for sparse observation
+    data assimilation, but can also be used to attend to gapless data like
+    topography.
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py]: Maybe specify that the sup3r latent space hypercube is query and obs/data-to-assimilate is key/value?

File: phygnn/layers/custom_layers.py [hunk 5]
```diff
@@ -87,6 +87,475 @@ def call(self, x):
...
+        embed_dim : int
+            Dimension of the positional encoding. This should match the
+            embed_dim of the attention block.
+        """
+        q_enc = self._pos_encoding(
+            x, patch_size=self.patch_size, dim_index=1, embed_dim=embed_dim
+        )
+        q_enc += self._pos_encoding(
+            x, patch_size=self.patch_size, dim_index=2, embed_dim=embed_dim
+        )
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py]: I think traditionally that the row/column positional encodings are done each for embed_dim//2 and then concatenated then added to the tensor x. Yes just confirmed this is at least how SatMAE does it.

File: phygnn/layers/custom_layers.py [hunk 6]
```diff
@@ -87,6 +87,475 @@ def call(self, x):
...
+        super().__init__(name=name)
+        self.rank = None
+        self.final_proj = None
+        self.out_proj = None
+        self.features = features or []
+        self.num_heads = num_heads
+        self.key_dim = key_dim
+        self.patch_size = patch_size
+        self.embed_dim = embed_dim
+        self.dropout = tf.keras.layers.Dropout(dropout)
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py]: I think that standard attention mechanisms including tf.keras.layers.MultiHeadAttention might include the "output" projections and dropout as part of the MHA class. You can confirm this by checking the MHA attribute "output_dense"

File: phygnn/layers/custom_layers.py [hunk 7]
```diff
@@ -87,6 +87,475 @@ def call(self, x):
...
+            embed_dim=embed_dim,
+        )
+        v = self._tokenize(
+            hi_res_feature,
+            patch_size=self.patch_size,
+            embed_dim=embed_dim,
+        )
+        return q, v
+
+    def encode(self, x, hi_res_feature, embed_dim=64):
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py]: I would call this "add_pos" or "add_pos_enc". Encode makes me think of a self-attention encoder transformer block.

**Review by grantbuster — CHANGES_REQUESTED** [review]
Generally i think it looks good but i kind of want to hear your plan for training positions vs. inference global/chunk positions. This is also pretty experimental and i just want to see how it goes. I've been having a tough time getting this to work haha.

**Comment by bnb32** [comment]
> Generally i think it looks good but i kind of want to hear your plan for training positions vs. inference global/chunk positions. This is also pretty experimental and i just want to see how it goes. I've been having a tough time getting this to work haha.

Yeah I thought about the inference issue a bit before opting for the relative encoding. I decided adding obs doesn't change the currently operative assumption that a fwp chunk includes everything that significantly influences the output (if not then we should increase padding).

File: phygnn/layers/custom_layers.py (hunk 3 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py]: If the key data wasn't sparse the patch size would be used for key data also. When it is sparse data the patch size is only used for query data and patch_size = 1 for key data.

File: phygnn/layers/custom_layers.py (hunk 6 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py]: Good call on dropout. I included the extra projection since you can change the attention shape with the output_shape arg but maybe that flexibility isn't worth it.

File: phygnn/layers/custom_layers.py (hunk 2 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py]: Yeah good call. Will rework this.

File: phygnn/layers/custom_layers.py (hunk 5 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py]: Hmm, then how would that change with additional depth/time encoding? Seems simplest to use embed_dim channels for each encoding.

File: phygnn/layers/custom_layers.py (hunk 1 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py]: Agreed, linspace is better.

Good points here. I hadn't thought about this potential issue on effective resolution. I'd expected that as long as fwp chunk shapes are similar size to batch shapes this will be fine, but maybe lat / lon / time features should actually be required inputs for this.

File: phygnn/layers/custom_layers.py [hunk 8]
```diff
@@ -87,6 +87,449 @@ def call(self, x):
...
+            Height, width, and depth of patches for value tokenization. Default
+            is 1 for pixel-wise tokenization. If value data is sparse this must
+            be 1. For gapless value data this can be > 1.
+        **kwargs : dict
+            Additional keyword arguments to pass to the MultiHeadAttention
+            layer.  This can include arguments like dropout for the attention
+            layer.
+        """
+
+        super().__init__(name=name)
```
[the remark this answers is not in the document]
Comment (bnb32) [reply phygnn/layers/custom_layers.py]: The doc string explicitly says these kwargs are only going to the MHA layer.

File: phygnn/layers/custom_layers.py [hunk 9]
```diff
@@ -87,6 +87,469 @@ def call(self, x):
...
+        self.num_heads = num_heads
+        self.key_dim = key_dim
+        self.embed_dim = embed_dim
+        self.q_patch_size = query_patch_size
+        self.v_patch_size = value_patch_size
+        self.attention = tf.keras.layers.MultiHeadAttention(
+            num_heads=self.num_heads,
+            key_dim=self.key_dim,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (bnb32) [reply phygnn/layers/custom_layers.py:766]: The doc string says that the kwargs only go to the MHA layer.

File: phygnn/layers/custom_layers.py [hunk 10]
```diff
@@ -87,6 +87,469 @@ def call(self, x):
...
+            f'shape: {input_shape}'
+        )
+        if self.rank not in {4, 5}:
+            logger.error(msg)
+            raise ValueError(msg)
+
+        self.final_proj = tf.keras.layers.Dense(input_shape[-1])
+
+        up_kwargs = {
+            'filters': input_shape[-1],
```
[the remark this answers is not in the document]
Comment (bnb32) [reply phygnn/layers/custom_layers.py]: I'm changing the signature to include both inputs explicitly, instead.

File: phygnn/layers/custom_layers.py [hunk 11]
```diff
@@ -87,6 +87,449 @@ def call(self, x):
...
+            amount of padding needed for each spatial dimension. Default is 1
+            for pixel-wise tokenization.
+        """
+        pads = [[0, 0]]  # batch
+        for i in range(1, len(x_shape) - 1):
+            rem = x_shape[i] % patch_size
+            pad = (patch_size - rem) % patch_size
+            pads.append([pad // 2, pad - pad // 2])
+        pads.append([0, 0])  # features
+        return pads
```
[the remark this answers is not in the document]
Comment (bnb32) [reply phygnn/layers/custom_layers.py:159]: x_shape comes from tf.shape(x) already.

**Comment by bnb32** [comment]
Closing - this is superseded by #73
